File: apps/api/app/schemas/earn.py
"""Pydantic schemas for Earn admin API。"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator


# ─────────────────────────────────────────────────────────
# Common
# ─────────────────────────────────────────────────────────


class FriendUserOption(BaseModel):
    """既有 user 的精簡資料,給 admin 從下拉選友。"""

    id: int
    email: str
    display_name: str | None
    earn_tier: str  # 'none' / 'internal' / 'friend' / 'commercial'


# ─────────────────────────────────────────────────────────
# Create / list
# ─────────────────────────────────────────────────────────


class CreateEarnAccountIn(BaseModel):
    user_id: int = Field(..., description="既有 user.id,且 earn_tier 應為 'none' 或 'friend'")

    # tier: 'friend' / 'internal'(Phase 1 不允許 'commercial')
    earn_tier: str = Field(default="friend", pattern=r"^(friend|internal)$")

    # custody_mode 預設 self(F-Phase 1 唯一支援)
    custody_mode: str = Field(default="self", pattern=r"^(self|platform)$")

    # Phase 1 預設 0(無抽成),未來可改為 500 / 1500
    perf_fee_bps: int = Field(default=0, ge=0, le=5000)

    can_quiver_operate: bool = Field(default=False)

    # Bitfinex 連線
    bitfinex_api_key: str = Field(..., min_length=20, max_length=200)
    bitfinex_api_secret: str = Field(..., min_length=20, max_length=200)
    bitfinex_permissions: str = Field(default="read", pattern=r"^(read|read\+funding-write)$")

    # EVM 地址
    evm_polygon_address: str | None = Field(
        default=None, pattern=r"^0x[0-9a-fA-F]{40}$"
    )
    evm_label: str | None = Field(default=None, max_length=64)

    # 備註
    notes: str | None = Field(default=None, max_length=2000)

    @field_validator("bitfinex_api_key", "bitfinex_api_secret")
    @classmethod
    def _strip_whitespace(cls, v: str) -> str:
        return v.strip()


class EarnAccountOut(BaseModel):
    id: int
    user_id: int
    user_email: str
    user_display_name: str | None
    earn_tier: str

    custody_mode: str
    perf_fee_bps: int
    can_quiver_operate: bool

    onboarded_by: int | None
    onboarded_by_email: str | None
    risk_acknowledged_at: datetime | None
    notes: str | None
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime

    has_active_bitfinex: bool
    bitfinex_permissions: str | None
    evm_addresses_count: int


class EarnAccountListOut(BaseModel):
    items: list[EarnAccountOut]
    total: int


# ─────────────────────────────────────────────────────────
# Detail (with positions)
# ─────────────────────────────────────────────────────────


class EvmAddressOut(BaseModel):
    id: int
    chain: str
    address: str
    is_platform_address: bool
    label: str | None
    created_at: datetime


class BitfinexConnectionOut(BaseModel):
    id: int
    is_platform_key: bool
    permissions: str
    has_key: bool  # True if encrypted_api_key present
    created_at: datetime
    revoked_at: datetime | None


class PositionSnapshotOut(BaseModel):
    snapshot_date: date
    bitfinex_funding_usdt: Decimal | None
    bitfinex_lent_usdt: Decimal | None
    bitfinex_daily_earned: Decimal | None
    aave_polygon_usdt: Decimal | None
    aave_daily_apr: Decimal | None
    total_usdt: Decimal | None


class EarnPipelinePositionOut(BaseModel):
    """F-Phase 3 Path A pipeline state — one row per deposit auto-lend cycle."""
    id: int
    status: str
    amount: Decimal
    currency: str
    onchain_tx_hash: str | None
    onchain_broadcast_at: datetime | None
    bitfinex_credited_at: datetime | None
    bitfinex_offer_id: int | None
    bitfinex_offer_submitted_at: datetime | None
    closed_at: datetime | None
    closed_reason: str | None
    last_error: str | None
    retry_count: int
    created_at: datetime


class EarnAccountDetailOut(EarnAccountOut):
    bitfinex_connections: list[BitfinexConnectionOut]
    evm_addresses: list[EvmAddressOut]
    recent_snapshots: list[PositionSnapshotOut]  # 最近 30 天
    # F-Phase 3 Path A:auto-lend pipeline positions (含 in-flight + recent closed)
    auto_lend_enabled: bool
    bitfinex_funding_address: str | None
    pipeline_positions: list[EarnPipelinePositionOut]


# ─────────────────────────────────────────────────────────
# Update
# ─────────────────────────────────────────────────────────


class UpdateEarnAccountIn(BaseModel):
    perf_fee_bps: int | None = Field(default=None, ge=0, le=5000)
    can_quiver_operate: bool | None = None
    notes: str | None = Field(default=None, max_length=2000)
    archived: bool | None = None


# ─────────────────────────────────────────────────────────
# Sync trigger result
# ─────────────────────────────────────────────────────────


class SyncResultOut(BaseModel):
    earn_account_id: int
    success: bool
    bitfinex_funding_usdt: Decimal | None = None
    bitfinex_lent_usdt: Decimal | None = None
    aave_polygon_usdt: Decimal | None = None
    total_usdt: Decimal | None = None
    error: str | None = None


# ─────────────────────────────────────────────────────────
# Cross-account APY ranking
# ─────────────────────────────────────────────────────────


class FriendApySummary(BaseModel):
    """每個 friend 過去 30 天的 APY 表現總結。"""

    earn_account_id: int
    user_email: str
    user_display_name: str | None
    total_usdt: Decimal | None
    avg_30d_apy_pct: Decimal | None  # 0-100
    bitfinex_share_pct: Decimal | None  # 0-100
    aave_share_pct: Decimal | None

File: apps/web/lib/api/earn.ts
/**
 * Earn admin API client — Friends Tooling F-Phase 1。
 */

import { apiFetch } from "@/lib/api";

export type EarnTier = "none" | "internal" | "friend" | "commercial";
export type CustodyMode = "self" | "platform";
export type BitfinexPermissions = "read" | "read+funding-write";

// ──── outputs ────

export interface FriendUserOption {
  id: number;
  email: string;
  display_name: string | null;
  earn_tier: EarnTier;
}

export interface EarnAccountOut {
  id: number;
  user_id: number;
  user_email: string;
  user_display_name: string | null;
  earn_tier: EarnTier;

  custody_mode: CustodyMode;
  perf_fee_bps: number;
  can_quiver_operate: boolean;

  onboarded_by: number | null;
  onboarded_by_email: string | null;
  risk_acknowledged_at: string | null;
  notes: string | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string;

  has_active_bitfinex: boolean;
  bitfinex_permissions: BitfinexPermissions | null;
  evm_addresses_count: number;
}

export interface EarnAccountListOut {
  items: EarnAccountOut[];
  total: number;
}

export interface BitfinexConnectionOut {
  id: number;
  is_platform_key: boolean;
  permissions: BitfinexPermissions;
  has_key: boolean;
  created_at: string;
  revoked_at: string | null;
}

export interface EvmAddressOut {
  id: number;
  chain: string;
  address: string;
  is_platform_address: boolean;
  label: string | null;
  created_at: string;
}

export interface PositionSnapshotOut {
  snapshot_date: string;
  bitfinex_funding_usdt: string | null;
  bitfinex_lent_usdt: string | null;
  bitfinex_daily_earned: string | null;
  aave_polygon_usdt: string | null;
  aave_daily_apr: string | null;
  total_usdt: string | null;
}

export interface EarnPipelinePositionOut {
  id: number;
  status: string;
  amount: string;
  currency: string;
  onchain_tx_hash: string | null;
  onchain_broadcast_at: string | null;
  bitfinex_credited_at: string | null;
  bitfinex_offer_id: number | null;
  bitfinex_offer_submitted_at: string | null;
  closed_at: string | null;
  closed_reason: string | null;
  last_error: string | null;
  retry_count: number;
  created_at: string;
}

export interface EarnAccountDetailOut extends EarnAccountOut {
  bitfinex_connections: BitfinexConnectionOut[];
  evm_addresses: EvmAddressOut[];
  recent_snapshots: PositionSnapshotOut[];
  // F-Phase 3 Path A
  auto_lend_enabled: boolean;
  bitfinex_funding_address: string | null;
  pipeline_positions: EarnPipelinePositionOut[];
}

export interface SyncResultOut {
  earn_account_id: number;
  success: boolean;
  bitfinex_funding_usdt: string | null;
  bitfinex_lent_usdt: string | null;
  aave_polygon_usdt: string | null;
  total_usdt: string | null;
  error: string | null;
}

export interface FriendApySummary {
  earn_account_id: number;
  user_email: string;
  user_display_name: string | null;
  total_usdt: string | null;
  avg_30d_apy_pct: string | null;
  bitfinex_share_pct: string | null;
  aave_share_pct: string | null;
}

// ──── inputs ────

export interface CreateEarnAccountIn {
  user_id: number;
  earn_tier: "friend" | "internal";
  custody_mode: CustodyMode;
  perf_fee_bps: number;
  can_quiver_operate: boolean;
  bitfinex_api_key: string;
  bitfinex_api_secret: string;
  bitfinex_permissions: BitfinexPermissions;
  evm_polygon_address?: string | null;
  evm_label?: string | null;
  notes?: string | null;
}

export interface UpdateEarnAccountIn {
  perf_fee_bps?: number;
  can_quiver_operate?: boolean;
  notes?: string;
  archived?: boolean;
}

// ──── client functions ────

export async function fetchEligibleUsers(): Promise<FriendUserOption[]> {
  return apiFetch<FriendUserOption[]>("/api/admin/earn/users");
}

export async function fetchEarnAccounts(opts: {
  includeArchived?: boolean;
} = {}): Promise<EarnAccountListOut> {
  const params = new URLSearchParams();
  if (opts.includeArchived) params.set("include_archived", "true");
  const qs = params.toString();
  return apiFetch<EarnAccountListOut>(
    `/api/admin/earn/accounts${qs ? `?${qs}` : ""}`,
  );
}

export async function fetchEarnAccountDetail(
  id: number,
): Promise<EarnAccountDetailOut> {
  return apiFetch<EarnAccountDetailOut>(`/api/admin/earn/accounts/${id}`);
}

export async function createEarnAccount(
  input: CreateEarnAccountIn,
): Promise<EarnAccountOut> {
  return apiFetch<EarnAccountOut>("/api/admin/earn/accounts", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function updateEarnAccount(
  id: number,
  input: UpdateEarnAccountIn,
): Promise<EarnAccountOut> {
  return apiFetch<EarnAccountOut>(`/api/admin/earn/accounts/${id}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export async function revokeBitfinexConnection(
  accountId: number,
  connectionId: number,
): Promise<void> {
  await apiFetch(
    `/api/admin/earn/accounts/${accountId}/connections/${connectionId}/revoke`,
    { method: "POST" },
  );
}

export async function syncEarnAccount(id: number): Promise<SyncResultOut> {
  return apiFetch<SyncResultOut>(`/api/admin/earn/accounts/${id}/sync`, {
    method: "POST",
  });
}

export async function syncAllEarnAccounts(): Promise<SyncResultOut[]> {
  return apiFetch<SyncResultOut[]>("/api/admin/earn/sync-all", {
    method: "POST",
  });
}

export async function fetchEarnRanking(): Promise<FriendApySummary[]> {
  return apiFetch<FriendApySummary[]>("/api/admin/earn/ranking");
}
